chain/chain/src/flat_storage_resharder.rs

```diff
@@ -396,8 +405,74 @@ impl FlatStorageResharder {
+        &self,
+        flat_store: &'a FlatStoreAdapter,
+        shard_uid: &ShardUId,
+        block_hash: &CryptoHash,
+    ) -> Result<Box<FlatStorageAndDeltaIter<'a>>, Error> {
+        let mut iter: Box<FlatStorageAndDeltaIter<'a>> = Box::new(
+            flat_store
+                .iter(*shard_uid)
+                // Get the flat storage iter and wrap the value in Optional::Some to
+                // match the delta iterator so that they can be chained.
+                .map_ok(|(key, value)| (key, Some(value)))
+                // Wrap the iterator's item into an Entry.
+                .map(|entry| FlatStorageAndDeltaIterItem::Entry(entry)),
+        );
+
+        // Get all the blocks from flat head to the wanted block hash.
+        let flat_storage = self
+            .runtime
+            .get_flat_storage_manager()
+            .get_flat_storage_for_shard(*shard_uid)
+            .expect("the flat storage undergoing resharding must exist!");
+        // Must reverse the result because we want ascending block heights.
+        let mut blocks_to_head = flat_storage.get_blocks_to_head(block_hash).map_err(|err| {
+            StorageError::StorageInconsistentState(format!(
+                "failed to find path of blocks to flat storage head ({err})"
+            ))
+        })?;
+        blocks_to_head.reverse();
+        debug!(target = "resharding", "flat storage blocks to head len = {}", blocks_to_head.len());
+
+        // Get all the delta iterators and wrap the items in Result to match the flat
+        // storage iter so that they can be chained.
+        for block in blocks_to_head {
+            let deltas = flat_store.get_delta(*shard_uid, block).map_err(|err| {
+                StorageError::StorageInconsistentState(format!(
+                    "can't retrieve deltas for flat storage at {block}/{shard_uid:?}({err})"
+                ))
+            })?;
+            let Some(deltas) = deltas else {
+                continue;
+            };
+            // Chain the iterators effectively adding a block worth of deltas.
+            // Before doing so insert a commit point to separate changes to the same key in different transactions.
+            iter = Box::new(iter.chain(iter::once(FlatStorageAndDeltaIterItem::CommitPoint)));
+            let deltas_iter = deltas.0.into_iter();
+            let deltas_iter = deltas_iter.map(|item| FlatStorageAndDeltaIterItem::Entry(Ok(item)));
+            iter = Box::new(iter.chain(deltas_iter));
+        }
+
+        Ok(iter)
+    }
+}
+
+/// Enum used to wrap the `Item` of iterators over flat storage contents or flat storage deltas. Its
+/// purpose is to insert a marker to force store commits during iteration over all entries. This is
+/// necessary because otherwise deltas might set again the value of a flat storage entry inside the
+/// same transaction.
+enum FlatStorageAndDeltaIterItem {
+    Entry(Result<(Vec<u8>, Option<FlatStateValue>), FlatStorageError>),
```

Comment:
nit: Is there any chance to remove the error from here? Would it be nicer?

Reply:
The only way I think would be to `map` the error into a `panic`, but that's.. brutal